evennia/contrib/game_systems/achievements/achievements.py

```diff
@@ -0,0 +1,370 @@
+
+
+def all_achievements():
+    """
+    Returns a dict of all achievements in the game.
+    """
+    # we do this to prevent accidental in-memory modification of reference data
+    return dict((key, dict(val)) for key, val in _load_achievements().items())
+
+
+def get_progress(achiever, key):
+    """
+    Retrieve the progress data on a particular achievement for a particular achiever.
+
+    Args:
+        achiever (Account or Character):  The entity tracking achievement progress.
+        key (str): The achievement key
+
+    Returns:
+        data (dict): The progress data
+    """
+    # split out the achievement attribute info
+    attr_key = _ACHIEVEMENT_ATTR[0]
+    attr_cat = _ACHIEVEMENT_ATTR[1] if len(_ACHIEVEMENT_ATTR) > 1 else None
+    if progress_data := achiever.attributes.get(attr_key, default={}, category=attr_cat):
+        # detach the data from the db to avoid data corruption and return the data
+        return progress_data.deserialize().get(key, {})
+    else:
+        # just return an empty dict
+        return {}
+
+
+def search_achievement(search_term):
+    """
+    Search for an achievement by name.
+
+    Args:
+        search_term (str):  The string to search for.
+
+    Returns:
+        results (dict):  A dict of key:data pairs of matching achievements.
+    """
+    if not (all_achievements := _load_achievements()):
+        # there are no achievements available, there's nothing to do
+        return {}
+    keys, names = zip(*((key, val["name"]) for key, val in all_achievements.items()))
+    indices = string_partial_matching(names, search_term)
+
+    return dict((keys[i], dict(all_achievements[keys[i]])) for i in indices)
+
+
+class CmdAchieve(MuxCommand):
+    """
+    view achievements
+
+    Usage:
+        achievements[/switches] [args]
+
+    Switches:
+        all          View all achievements, including locked ones.
+        completed    View achievements you've completed.
+        progress     View achievements you have partially completed
+
+    Check your achievement statuses or browse the list. Providing a command argument
+    will search all your currently unlocked achievements for matches, and the switches
+    will filter the list to something other than "all unlocked". Combining a command
+    argument with a switch will search only in that list.
+
+    Examples:
+        achievements apples
+        achievements/all
+        achievements/progress rats
+    """
+
+    key = "achievements"
+    aliases = (
+        "achievement",
+        "achieve",
+    )
+    switch_options = ("progress", "completed", "done", "all")
+
+    def format_achievement(self, achievement_data):
+        """
+        Formats the raw achievement data for display.
+
+        Args:
+            achievement_data (dict):  The data to format.
+
+        Returns
+            str: The display string to be sent to the caller.
+
+        """
+        template = """\
+|w{name}|n
+{desc}
+{status}
+""".rstrip()
```

Comment:
Why embed this inside the `format_achievement` method? Put it on the class instead - then if people don't want to change much, they can just do 

```py
class CmdMyAchieve(CmdAchieve): 
    template = """ my custom style """ 
```

Reply:
I had it that way originally, but I realized that most, if not all, people who want to override the template will want to change the strings being *put into* the template as well, making it fairly useless to override just the template. I'll split it out anyway though.

Reply:
While you are probably right in practice, I just know that _someone_ will eventually ask why this simple override is not possible down the line :wink: Breaking it out costs nothing, so best to go for max flexibility here.